### PR TITLE
Valid TTH check & proper spawn times. --Fast-scan updates

### PR DESCRIPTION
**Pull request by mew905**
## Description
Detects long spawns (but not hidden time), checks TTH value, if it's a good value (you catch it within 90 seconds of despawn) it flags it in the database, if not, it checks the spawn_id against the database to see if it's been scanned with a valid timer in the past, grabs the most recent one, calculates the difference between then and now, and provides you with a future spawn time. Also happens to detect long spawns.

Also includes a database pruner activated with -di flag (--delete-invalid) that will prune all pokemon with invalid despawn time AFTER the map's assumed 15 minute despawn.

## Motivation and Context
Prepping for spawn scan, still useless without a proper speed limiter, but I dont like being told something's not possible.... plus user complaints of pokemon being on the map, but not on the game, were annoying.

## How Has This Been Tested?
On my own map, see screenshot below

## Screenshots (if appropriate):
image
image

## Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mew905** [comment]
throws a list out of range exception, Trying to find the cause but not being good at python makes me even worse at debugging

File: pogom/models.py [hunk 1]
```diff
@@ -802,9 +814,27 @@ def parse_map(args, map_dict, step_location, db_update_queue, wh_update_queue, a
...
+                query = (Pokemon
+                         .select(Pokemon.disappear_time, Pokemon.last_modified)  # we only need the times
+                         .where((Pokemon.spawnpoint_id == p['spawn_point_id']) & (Pokemon.disappear_time < datetime.utcnow()) & (Pokemon.valid == 1))  # grab matching spawn point ID with a valid timer
+                         .order_by(Pokemon.disappear_time.desc())  # sort by most recent
+                         .limit(1)  # we only want one
+                         .dicts())
+
+                # make this a datetime object now
+                nptime = datetime.utcfromtimestamp(p['last_modified_timestamp_ms'] / 1000.0)
+                if query.count > 0:
```
Comment (mew905) [inline pogom/models.py]: possible source of list out of range exception?

Comment (mew905) [inline pogom/models.py]: well this is awkward. Apparently I cant Git

**Comment by mew905** [comment]
"Disappear time unknown" doesn't work now. Thanks, Travis >.> I wish I knew how to js

**Comment by omgms** [comment]
It broke again hahaha

File: pogom/utils.py [hunk 2]
```diff
@@ -154,6 +154,9 @@ def get_args():
     parser.add_argument('-pd', '--purge-data',
                         help='Clear pokemon from database this many hours after they disappear \
                         (0 to disable)', type=int, default=0)
+    parser.add_argument('-di', '--delete_invalid',
```
Comment (sLoPPydrive) [inline pogom/utils.py]: Better call it "delete-invalid" - use dash instead of underscore.

**Comment by sLoPPydrive** [comment]
I'm having the issue that after I refresh my browser no Pokemon will be shown. This might be a UI only problem as Pokemon are being found and webhooks being triggered.

Maybe it is a problem that I use --delete_invalid for both the map process (no scanner) and the scanner processes (no web server)?

**Comment by sLoPPydrive** [comment]
Ah well, and no Pokemon details are shown. I have encounters enabled (for most Pokemon types).
image

**Comment by Extrememist** [comment]
Have you tried it without -di ?

**Comment by midadami** [comment]
It seems it works fine without -di, I am starting to get some 30min spawns. I know the timers might be wrong and actually checking its dificult.

**Comment by sLoPPydrive** [comment]
Giving it another try without `-di`. So far the Pokemon won't disappear when I refresh the browser page, yay! But clicking on a Pokemon still doesn't show any details. After checking out the pull request I did a `pip install -r requirements.txt --upgrade` (not needed for this PR I think) and `npm install`. What could be the cause? :-/

**Comment by Barloklol** [comment]
It's some error in the map.js file, trying to figure it out

**Comment by Extrememist** [comment]
all i know is i didn't apply the rest of the update so i cannot help much further. :D

**Comment by Barloklol** [comment]
The code cleanup fucked it, use the map.js code from this commit https://github.com/PokemonGoMap/PokemonGo-Map/pull/1527/commits/a3bcec961ac3db0d96050135993e808cd0174c0f

**Comment by KoertJanssens** [comment]
Shouldn't we delete 
`if (b64encode(str(p['encounter_id'])), p['spawn_point_id']) in encountered_pokemon:
                # If pokemon has been encountered before dont process it.
                skipped += 1
                continue`
from _models.py_ in order to update the valid time if we scan a pokemon who is already in the DB?
Or better: skip all the valid Pokemon?

**Comment by sLoPPydrive** [comment]
I just looked at `map.js` and learned about the backticks in JS (template literals). What I still don't get is the conditional concatenation of template literals like this:
```
var contentstring = `
	<div>...<div>`
	if (valid === 1) {
	  `....`
	} else {
	  `Disappear time unknown`
	}`</div>`
```
Is this even proper JS syntax? Couldn't find an example anywhere...

**Comment by midadami** [comment]
Im getting some 1h30 min spawn haha. Not sure if thats right.

**Comment by quixfz** [comment]
great idea, looking forward to a working version.

but, this one doesnt work. problems i see:
- i installed it and getting 1h30 spawns
- see KoertJanssens's comment. currently only spawns scanned (for the first time) in the last 90 seconds (before hidden) get a valid=1 flag.

- _last_modified returned by api is different then last_modified from db (my understanding at least). api time is when it appeared. db time is when it was last modified (currently == appear time), but when updating disappear_time, last_modified should be updated and that should lead to it being sent to the web gui. it should be possible the generate the new disappear_time with the old one. (without using last modified) PS: ignore all that, doenst matter when point above gets fixed._

my understanding of this: run hex-scan (at least an hour) with faster then 90seconds scan time to get a valid time for each spawnpoint. after that you can slow down the scan.

i'm currently running dozens of st3 sd11 scanner, with 3 accounts each. this should take under 90 seconds.

**Comment by Barloklol** [comment]
I removed the line as suggested by KoertJanssen  and currently testing in all my cities. I noticed 1h+ timers as well, not sure if the others are correct. Told my users to report back to me how the "safe" timers are working out, so we'll see.

**Comment by mew905** [comment]
yeah, not sure if map.js works anymore, works on my map as you can see from the screenshots, but travis had me change the if valid === 0 logic to stop the check errors and I dont know how to fix the new logic, home assistance in this area would be well appreciated. Only changes are under pokemonLabel logic.

1h+ timers are definitely an error, but not from my 'fix'. last_modified goes into the database as default utctimenow() when it is upserted, if it exceeds the calculated disappear_time when it goes into the database (could happen because of database locks or slow connections) then the spawn_check logic produces a 1+hour spawn. I did an update in the -di logic that cleans up those times, as they're invalid anyway (how did we find a despawned pokemon? we didn't), so I changed it so it drops the UTC equivalent last_modified_ms into the database instead of utctimenow(). This should add it to the database as "we found it at this time" instead of "we inserted it at this time".

**Comment by slideup-benni** [comment]
I think to delete pokemon from the database to get correct spawn points is not a good idea. You will also remove them from the stats-history when you delete them. A better way is to move spawnpoints to a separate table like here 

https://github.com/PokemonGoMap/PokemonGo-Map/pull/1494

Anyway. You've did a good job! thank you

File: static/js/map.js [hunk 3]
```diff
@@ -362,35 +362,56 @@ function pokemonLabel (name, rarity, types, disappearTime, id, latitude, longitu
...
-	</div>
-	${details}
-	<div>
-	<a href='javascript:excludePokemon(${id})'>Exclude</a>&nbsp;&nbsp
-	<a href='javascript:notifyAboutPokemon(${id})'>Notify</a>&nbsp;&nbsp
-	<a href='javascript:void(0);' onclick='javascript:openMapDirections(${latitude},${longitude});' title='View in Maps'>Get directions</a>
-	</div>`
-  return contentstring
+  if (valid > 0) {
+	var contentstring = `
```
Comment (nullpixel1) [inline static/js/map.js]: You should indent with 4 spaces here

File: static/js/map.js [hunk 4]
```diff
@@ -362,35 +362,56 @@ function pokemonLabel (name, rarity, types, disappearTime, id, latitude, longitu
...
+		<div>
+		Location: ${latitude.toFixed(6)}, ${longitude.toFixed(7)}
+		</div>
+		${details}
+		<div>
+		<a href='javascript:excludePokemon(${id})'>Exclude</a>&nbsp;&nbsp
+		<a href='javascript:notifyAboutPokemon(${id})'>Notify</a>&nbsp;&nbsp
+		<a href='javascript:void(0);' onclick='javascript:openMapDirections(${latitude},${longitude});' title='View in Maps'>Get directions</a>
+		</div>`
+  }  else {
```
Comment (nullpixel1) [inline static/js/map.js]: Multiple spaces before else

File: static/js/map.js [hunk 5]
```diff
@@ -362,35 +362,56 @@ function pokemonLabel (name, rarity, types, disappearTime, id, latitude, longitu
...
+		var contentstring = `
```
Comment (nullpixel1) [inline static/js/map.js]: Travis expects 4 spaces, but there are 2 tabs instead

Comment (nullpixel1) [inline static/js/map.js]: `contentstring` is already defined, you redefined it

File: static/js/map.js [hunk 6]
```diff
@@ -362,35 +362,56 @@ function pokemonLabel (name, rarity, types, disappearTime, id, latitude, longitu
...
+		<div>
+		Location: ${latitude.toFixed(6)}, ${longitude.toFixed(7)}
+		</div>
+		${details}
+		<div>
+		<a href='javascript:excludePokemon(${id})'>Exclude</a>&nbsp;&nbsp
+		<a href='javascript:notifyAboutPokemon(${id})'>Notify</a>&nbsp;&nbsp
+		<a href='javascript:void(0);' onclick='javascript:openMapDirections(${latitude},${longitude});' title='View in Maps'>Get directions</a>
+		</div>`
+	}  return contentstring
```
Comment (nullpixel1) [inline static/js/map.js]: Travis wants two spaces rather than a tab

Comment (nullpixel1) [inline static/js/map.js]: Multiple spaces before return

**Comment by mew905** [comment]
No longer deletes active, invalid pokemon. Didn't know upsert meant "update if insert has matching unique ID". Now updates existing invalid pokemon with newly scanned pokemon for those who happen to scan the same pokemon multiple times before despawn.

-di is only a database cleanup tool that erased despawned invalid pokemon, as well as pokemon whose insertion time was after despawn time, and is no longer necessary with the changes in models.py.

File: static/js/map.js [hunk 7]
```diff
@@ -362,12 +362,11 @@ function pokemonLabel (name, rarity, types, disappearTime, id, latitude, longitu
...
       `
   }
-  var dstime = ''
 
   if (valid > 0) {
-    dstime = `<div>Disappears at ${pad(disappearDate.getHours())}:${pad(disappearDate.getMinutes())}:${pad(disappearDate.getSeconds())}<span class='label-countdown' disappears-at='${disappearTime}'>(00m00s)</span></div>`
+    var dstime = `<div>Disappears at ${pad(disappearDate.getHours())}:${pad(disappearDate.getMinutes())}:${pad(disappearDate.getSeconds())}<span class='label-countdown' disappears-at='${disappearTime}'>(00m00s)</span></div>`
   } else {
-    dstime = `<div>Disappear time unknown</div>`
+    var dstime = `<div>Disappear time unknown</div>`
```
Comment (nullpixel1) [inline static/js/map.js]: Change to ```dstime = '<div>Disappear time unknown</div>' ```

Comment (nullpixel1) [inline static/js/map.js]: Travis is asking for single quotes, which are `''` these, so mayve

**Comment by Arengo** [comment]
PR#1530 and this should collaborate.

**Comment by mew905** [comment]
@arengo we actually discussed the logic before this came about. He was working on his but because he was busy and couldnt test until later i decided to make an attempt of my own. I had no idea the approaches would be so different. Because im not familiar with what python code uses what CPU cycles i have no idea whose implementation is more efficient but as far as i can tell, both work equally well.

**Comment by tokar86a** [comment]
That sad. But which going to be implemented in the map? Should be nice to see a fix to the spawn problem.

**Comment by houkop** [comment]
+1 to tokar86a, May of us can't decide whith one is better and would be in implemented.

**Comment by Skorfius** [comment]
has anyone else noticed this pr breaks Pokemon auto updating on the map, without manually refreshing the page?

**Comment by jackey1510** [comment]
Yes, sometimes it will not show some of the Pokémon spawned lately. Need to manually refresh the browser.

**Comment by sLoPPydrive** [comment]
Yep, same here. Some Pokémon are missing and are only shown after manual refresh.

**Comment by mew905** [comment]
Yes, its an issue i brought up in #pr as my map doesnt do it, and barlok said his did, but the latest version fixes it for him. Clearly thats not the case for everyone.

**Comment by Barloklol** [comment]
It's still a problem for some of my users, but for me it seems fixed. At least it refreshes now, but apparently it's missing some pokemon here and there until you fully refresh. Not really sure what's causing this because it seems fine when I check out the packets in Chromes network details.

**Comment by sLoPPydrive** [comment]
I bet it's the DB query to fetch only the new Pokemon for the given viewport (if such a query exists). Its conditions will most probably leave some Pokemon out.

**Comment by AdrienNguyenWorldline** [comment]
UPDATE pokevislyon.pokemon SET valid = 1 WHERE TIMESTAMPDIFF(SECOND,last_modified,disappear_time) < 90; 

Should be executed when updating database

**Comment by Obihoernchen** [comment]
I think this can be closed. We have -speed now and there are a lot of merge conflicts with last update ~3 month ago.

**Comment by sebastienvercammen** [comment]
Closing old PR, we've had SpeedScan for a while now.

**Pull request closed without merging**